index.js

```diff
@@ -1,96 +1,98 @@
 var path = require('path')
 var walk = require('fs-walk')
-var fs = require('fs')
-var mkdirp = require('mkdirp')
 
-function noop () {}
+var noop = function () {}
 
-function BlobStore (dir, opts) {
-  if (!(this instanceof BlobStore)) return new BlobStore(dir, opts)
+// Random (but stable) postfix used to id tmp files (so they aren't listed)
+var TMP_POSTFIX = '.tmp-gxqIdUqEoqo'
 
-  // TODO: expose whether to use subdirs opt
-  // TODO: expose subdir prefix length opt
-  // TODO: expose whether to use a 'staging' subdir
+function murmurhex () {
+  var hash = MurmurHash3('')
+  for (var ii = 0; ii < arguments.length; ++ii) {
+    hash.hash('' + arguments[ii])
+  }
+  return hash.result()
+}
 
-  this._dir = dir
-  this._stores = {}
+var invocations = 0
+function getTmpname (filename) {
+  return filename + '.' + murmurhex(__filename, process.pid, ++invocations) + TMP_POSTFIX
 }
 
-BlobStore.prototype._getStore = function (subdir) {
-  if (!this._stores[subdir]) {
-    this._stores[subdir] = Store(path.join(this._dir, subdir))
-  }
-  return this._stores[subdir]
+function BlobStore (opts) {
+  if (!(this instanceof BlobStore)) return new BlobStore(opts)
+  if (typeof opts === 'string') opts = {path: opts}
+  this.subDirPrefixLen = opts.subDirPrefixLen || 7
+  this.path = opts.path
+  AtomicStore.call(this, opts)
 }
 
-BlobStore.prototype._list = function (cb) {
-  var names = []
-  walk.files(this._dir, function (basedir, filename, stat, next) {
-    if (!basedir.endsWith('staging')) names.push(filename)
-    next()
-  }, function (err) {
-    if (err && err.code === 'ENOENT') cb(null, [])
-    else cb(err, names)
+inherits(BlobStore, AtomicStore)
+
+BlobStore.prototype.createWriteStream = function (opts, cb) {
+  if (typeof cb !== 'function') cb = noop
+  if (typeof opts === 'string') opts = {key: opts}
+  if (opts.name && !opts.key) opts.key = opts.name
+  var originalKey = opts.key
+  opts.getTmpname = getTmpname
+  opts.key = this._insertSubDirPrefix(opts.key)
+  return AtomicStore.prototype.createWriteStream.call(this, opts, function (err, metadata) {
+    metadata.key = originalKey
+    cb(err, metadata)
   })
 }
 
-BlobStore.prototype.createReadStream = function (opts) {
-  var name = typeof opts === 'string' ? opts : opts.key
-  var subdir = filenamePrefix(name, 7)
-  var store = this._getStore(subdir)
-  return store.createReadStream(opts)
+BlobStore.prototype.createReadStream = function (key) {
+  if (key && typeof key === 'object') return this.createReadStream(key.key)
+  key = this._insertSubDirPrefix(key)
+  return AtomicStore.prototype.createReadStream.call(this, key)
 }
 
-BlobStore.prototype.exists = function (opts, done) {
-  var name = typeof opts === 'string' ? opts : opts.key
-  var subdir = filenamePrefix(name, 7)
-  var store = this._getStore(subdir)
-  return store.exists(opts, done)
+BlobStore.prototype.exists = function (opts, cb) {
+  if (typeof opts === 'string') opts = {key: opts}
```

Comment:
`if (typeof cb !== 'function') cb = noop` (or fail if not given)

Reply:
This is handled by atomic-fs-blob-store